src/Extensions/AddToCampaignExtension.php

```diff
@@ -0,0 +1,128 @@
+     */
+    public function addtocampaign(array $data, Form $form): HTTPResponse
+    {
+        $owner = $this->getOwner();
+        $id = $data['ID'];
+        $dataClass = $this->getManagedDataClass();
+        $record = $dataClass::get()->byID($id);
+
+        $handler = AddToCampaignHandler::create($owner, $record, 'addToCampaignForm');
+        $response = $handler->addToCampaign($record, $data);
+        $message = $response->getBody();
+        if (empty($message)) {
+            return $response;
+        }
+
+        // Send extra "message" data with schema response
+        $extraData = ['message' => $message];
+        $schemaId = Controller::join_links($owner->Link('schema/addToCampaignForm'), $id);
+        return $owner->getSchemaResponse($schemaId, $form, null, $extraData);
+    }
+
+    /**
+     * Url handler for add to campaign form
+     */
+    public function addToCampaignForm(HTTPRequest $request): ?Form
+    {
+        // Get ID either from posted back value, or url parameter
+        $id = $request->param('ID') ?: $request->postVar('ID');
+        return $this->getAddToCampaignForm((int) $id);
+    }
+
+    /**
+     * Get the form for adding an item to a campaign
+     */
+    public function getAddToCampaignForm(int $id): ?Form
+    {
+        $owner = $this->getOwner();
+        // Get record-specific fields
+        $dataClass = $this->getManagedDataClass();
+        $record = $dataClass::get()->byID($id);
+
+        if (!$record) {
+            $owner->jsonError(404, _t(
+                __CLASS__.'.ErrorNotFound',
+                "That {Type} couldn't be found",
+                ['Type' => $dataClass::singleton()->i18n_singular_name()]
+            ));
+            return null;
+        }
+        if (!$record->canView()) {
+            $owner->jsonError(403, _t(
+                __CLASS__.'.ErrorItemPermissionDenied',
+                "You don't have the necessary permissions to modify {ObjectTitle}",
+                ['ObjectTitle' => $record->i18n_singular_name()]
+            ));
+            return null;
+        }
+
+        $handler = AddToCampaignHandler::create($owner, $record, 'addToCampaignForm');
+        $form = $handler->Form($record);
+
+        $form->setValidationResponseCallback(function (ValidationResult $errors) use ($owner, $form, $id) {
+            $schemaId = Controller::join_links($owner->Link('schema/addToCampaignForm'), $id);
+            return $owner->getSchemaResponse($schemaId, $form, $errors);
+        });
+
+        return $form;
+    }
+
+    /**
+     * Get the managed data class for this extension
+     */
+    private function getManagedDataClass(): string
+    {
+        $owner = $this->getOwner();
+        $property = 'campaign_admin_managed_data_class';
+        $dataClass = $owner->config()->get($property);
+        if (!is_a($dataClass, DataObject::class, true)) {
+            throw new RuntimeException("$owner.$property is not a DataObject subclass.");
```

Comment:
`$owner` here should not be trusted to evaluate to a class name. Use `get_class($owner)`

Reply:
Also, if it needs to explicitly be a subclass (i.e. can't be `DataObject` itself), use `is_subclass_of()` - but if `DataObject` is allowed, update the message to "is not DataObject or a subclass" or similar.

Reply:
Updated